language/vm/src/normalized.rs

```diff
@@ -0,0 +1,288 @@
+                            t.into_type_tag().expect(
+                                "Invariant violation: struct type argument contains reference",
+                            )
+                        })
+                        .collect(),
+                }),
+                TypeParameter(_) => unreachable!(),
+            }
+        } else {
+            return None;
+        })
+    }
+}
+
+impl Field {
+    /// Create a `Field` for `FieldDefinition` `f` in module `m`.
+    pub fn new(m: &CompiledModule, f: &FieldDefinition) -> Self {
+        Field {
+            name: m.identifier_at(f.name).to_owned(),
+            type_: Type::new(m, &f.signature.0),
+        }
+    }
+}
+
+impl Struct {
+    /// Create a `Struct` for `StructDefinition` `def` in module `m`. Panics if `def` is a
+    /// a native struct definition.
+    pub fn new(m: &CompiledModule, def: &StructDefinition) -> Self {
+        let handle = m.struct_handle_at(def.struct_handle);
+        let fields = match &def.field_information {
+            StructFieldInformation::Native => panic!("Can't extract  for native struct"),
+            StructFieldInformation::Declared(fields) => {
+                fields.iter().map(|f| Field::new(m, f)).collect()
+            }
+        };
+        Struct {
+            name: m.identifier_at(handle.name).to_owned(),
+            kind: Kind::from_bool(handle.is_nominal_resource),
+            type_parameters: handle
+                .type_parameters
+                .iter()
+                .map(|k| Kind::new(k))
+                .collect(),
+            fields,
+        }
+    }
+}
+
+impl FunctionSignature {
+    /// Create a `FunctionSignature` for `FunctionHandle` `f` in module `m`.
+    pub fn new(m: &CompiledModule, f: &FunctionHandle) -> Self {
+        FunctionSignature {
+            name: m.identifier_at(f.name).to_owned(),
+            type_parameters: f.type_parameters.iter().map(|k| Kind::new(k)).collect(),
+            formals: m
+                .signature_at(f.parameters)
+                .0
+                .iter()
+                .map(|s| Type::new(m, s))
+                .collect(),
+            ret: m
+                .signature_at(f.return_)
+                .0
+                .iter()
+                .map(|s| Type::new(m, s))
+                .collect(),
+        }
+    }
+}
+
+impl Kind {
+    /// Create a `Kind` from a bytecode file format `Kind`.
+    pub fn new(k: &file_format::Kind) -> Self {
+        match k {
+            file_format::Kind::All => Kind::All,
+            file_format::Kind::Resource => Kind::Resource,
+            file_format::Kind::Copyable => Kind::Copyable,
+        }
+    }
+
+    /// Construct a `Kind` from the `is_nominal_resource` bool used in the bytecode file format's
+    /// `StructDefinition`
+    pub fn from_bool(is_nominal_resource: bool) -> Self {
+        if is_nominal_resource {
+            Kind::Resource
+        } else {
+            Kind::Copyable
+        }
+    }
```

Comment:
My thoughts is to remove this as it feels misleading, but I could be wrong. 

The point I'm trying to make is that `is_nominal_resource` implies `Resource` but `Resource` does not imply `is_nominal_resource`

Reply:
Fair point; doesn't seem good to have this as a public API. I will make it private or inline it in the one place that it's used.